File: app/Services/ContentExtractorService.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Log;
use GuzzleHttp\Client;
use GuzzleHttp\Promise;
use Symfony\Component\DomCrawler\Crawler;

class ContentExtractorService
{
    protected $httpClient;

    // Noise elements yang perlu dihapus
    protected $noiseSelectors = [
        'script', 'style', 'nav', 'footer', 'header',
        '.sidebar', '.advertisement', '.ad', '.ads',
        '.cookie-banner', '.popup', '.modal',
        '.social-share', '.related-posts', '.comments',
        '#sidebar', '#footer', '#header', '#nav',
        '.widget', '.newsletter', '.subscription',
    ];

    public function __construct()
    {
        $this->httpClient = new Client([
            'timeout' => 30,
            'verify' => false, // Disable SSL verification for dev
            'headers' => [
                'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
                'Accept' => 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
                'Accept-Language' => 'en-US,en;q=0.9,id;q=0.8',
            ],
        ]);
    }

    /**
     * Fetch dan extract konten dari satu URL
     */
    public function extract(string $url): array
    {
        try {
            $response = $this->httpClient->get($url);
            $html = $response->getBody()->getContents();
            $statusCode = $response->getStatusCode();

            if ($statusCode !== 200) {
                return $this->errorResult($url, "HTTP $statusCode");
            }

            return $this->parseContent($html, $url);
        } catch (\Exception $e) {
            Log::error("Content extraction error for $url: " . $e->getMessage());
            return $this->errorResult($url, $e->getMessage());
        }
    }

    /**
     * Fetch multiple URLs secara parallel
     */
    public function extractMultiple(array $urls, int $concurrency = 5): array
    {
        $promises = [];
        $results = [];

        foreach ($urls as $url) {
            $promises[$url] = $this->httpClient->getAsync($url)->then(
                function ($response) use ($url) {
                    $html = $response->getBody()->getContents();
                    return $this->parseContent($html, $url);
                },
                function ($exception) use ($url) {
                    return $this->errorResult($url, $exception->getMessage());
                }
            );
        }

        // Execute promises dengan concurrency limit
        $settled = Promise\Utils::settle($promises)->wait();

        foreach ($settled as $url => $result) {
            if ($result['state'] === 'fulfilled') {
                $results[] = $result['value'];
            } else {
                $results[] = $this->errorResult($url, 'Request failed');
            }
        }

        return $results;
    }

    /**
     * Parse HTML dan extract konten utama
     */
    protected function parseContent(string $html, string $url): array
    {
        $crawler = new Crawler($html, $url);

        // Extract metadata
        $title = $this->extractTitle($crawler);
        $description = $this->extractMetaDescription($crawler);
        $author = $this->extractAuthor($crawler);
        $publishDate = $this->extractDate($crawler);
        $siteName = $this->extractSiteName($crawler, $url);

        // Extract main content (readability mode)
        $mainContent = $this->extractMainContent($crawler);

        // Extract images
        $images = $this->extractImages($crawler, $url);

        // Hitung word count
        $wordCount = str_word_count($mainContent);

        return [
            'url' => $url,
            'title' => $title,
            'description' => $description,
            'author' => $author,
            'publish_date' => $publishDate,
            'site_name' => $siteName,
            'content' => $mainContent,
            'word_count' => $wordCount,
            'images' => $images,
            'success' => true,
        ];
    }

    /**
     * Extract title dari halaman
     */
    protected function extractTitle(Crawler $crawler): string
    {
        // Priority: og:title > h1 > <title>
        try {
            // og:title
            $ogTitle = $crawler->filter('meta[property="og:title"]');
            if ($ogTitle->count() > 0) {
                return trim($ogTitle->attr('content'));
            }

            // h1
            $h1 = $crawler->filter('h1');
            if ($h1->count() > 0) {
                return trim($h1->first()->text());
            }

            // <title>
            $title = $crawler->filter('title');
            if ($title->count() > 0) {
                return trim($title->text());
            }
        } catch (\Exception $e) {
            // ignore
        }

        return 'Untitled';
    }

    /**
     * Extract meta description
     */
    protected function extractMetaDescription(Crawler $crawler): string
    {
        try {
            // og:description
            $ogDesc = $crawler->filter('meta[property="og:description"]');
            if ($ogDesc->count() > 0) {
                return trim($ogDesc->attr('content'));
            }

            // meta description
            $metaDesc = $crawler->filter('meta[name="description"]');
            if ($metaDesc->count() > 0) {
                return trim($metaDesc->attr('content'));
            }
        } catch (\Exception $e) {
            // ignore
        }

        return '';
    }

    /**
     * Extract author
     */
    protected function extractAuthor(Crawler $crawler): string
    {
        try {
            // meta author
            $metaAuthor = $crawler->filter('meta[name="author"]');
            if ($metaAuthor->count() > 0) {
                return trim($metaAuthor->attr('content'));
            }

            // article:author
            $articleAuthor = $crawler->filter('meta[property="article:author"]');
            if ($articleAuthor->count() > 0) {
                return trim($articleAuthor->attr('content'));
            }

            // Schema.org author
            $schemaAuthor = $crawler->filter('[itemprop="author"]');
            if ($schemaAuthor->count() > 0) {
                return trim($schemaAuthor->first()->text());
            }

            // Common class names
            $authorSelectors = ['.author', '.byline', '.writer', '[rel="author"]'];
            foreach ($authorSelectors as $selector) {
                try {
                    $el = $crawler->filter($selector);
                    if ($el->count() > 0) {
                        return trim($el->first()->text());
                    }
                } catch (\Exception $e) {
                    continue;
                }
            }
        } catch (\Exception $e) {
            // ignore
        }

        return '';
    }

    /**
     * Extract publish date
     */
    protected function extractDate(Crawler $crawler): string
    {
        try {
            // article:published_time
            $articleDate = $crawler->filter('meta[property="article:published_time"]');
            if ($articleDate->count() > 0) {
                return trim($articleDate->attr('content'));
            }

            // Schema.org datePublished
            $schemaDate = $crawler->filter('[itemprop="datePublished"]');
            if ($schemaDate->count() > 0) {
                return trim($schemaDate->attr('content') ?? $schemaDate->text());
            }

            // time element
            $timeEl = $crawler->filter('time');
            if ($timeEl->count() > 0) {
                return trim($timeEl->first()->attr('datetime') ?? $timeEl->first()->text());
            }
        } catch (\Exception $e) {
            // ignore
        }

        return '';
    }

    /**
     * Extract site name
     */
    protected function extractSiteName(Crawler $crawler, string $url): string
    {
        try {
            // og:site_name
            $ogSite = $crawler->filter('meta[property="og:site_name"]');
            if ($ogSite->count() > 0) {
                return trim($ogSite->attr('content'));
            }
        } catch (\Exception $e) {
            // ignore
        }

        return parse_url($url, PHP_URL_HOST) ?? '';
    }

    /**
     * Extract main content (readability mode)
     */
    protected function extractMainContent(Crawler $crawler): string
    {
        try {
            // Clone crawler supaya gak mengubah original
            $contentCrawler = clone $crawler;

            // Hapus noise elements
            foreach ($this->noiseSelectors as $selector) {
                try {
                    $contentCrawler->filter($selector)->each(function (Crawler $node) {
                        // Skip kalau gak ada parent
                    });
                } catch (\Exception $e) {
                    continue;
                }
            }

            // Coba extract dari <article> dulu
            $articles = $contentCrawler->filter('article');
            if ($articles->count() > 0) {
                $text = $this->cleanText($articles->first()->text());
                if (strlen($text) > 200) {
                    return $text;
                }
            }

            // Coba dari .content, .post-content, .article-content, .entry-content
            $contentSelectors = [
                '.post-content', '.article-content', '.entry-content',
                '.content', '.main-content', '#content', '#main',
                '[role="main"]', 'main',
            ];

            foreach ($contentSelectors as $selector) {
                try {
                    $el = $contentCrawler->filter($selector);
                    if ($el->count() > 0) {
                        $text = $this->cleanText($el->first()->text());
                        if (strlen($text) > 200) {
                            return $text;
                        }
                    }
                } catch (\Exception $e) {
                    continue;
                }
            }

            // Fallback: ambil semua <p> tags
            $paragraphs = $contentCrawler->filter('p');
            $text = '';
            $paragraphs->each(function (Crawler $p) use (&$text) {
                $pText = trim($p->text());
                if (strlen($pText) > 30) {
                    $text .= $pText . "\n\n";
                }
            });

            $text = $this->cleanText($text);
            if (strlen($text) > 100) {
                return $text;
            }

            // Last fallback: ambil body text
            $body = $contentCrawler->filter('body');
            if ($body->count() > 0) {
                return $this->cleanText($body->first()->text());
            }

            return '';
        } catch (\Exception $e) {
            Log::error("Content extraction error: " . $e->getMessage());
            return '';
        }
    }

    /**
     * Clean extracted text
     */
    protected function cleanText(string $text): string
    {
        // Remove extra whitespace
        $text = preg_replace('/\s+/', ' ', $text);
        // Remove multiple newlines
        $text = preg_replace('/\n{3,}/', "\n\n", $text);
        // Trim
        $text = trim($text);

        return $text;
    }

    /**
     * Extract images dari halaman
     */
    protected function extractImages(Crawler $crawler, string $baseUrl): array
    {
        $images = [];

        try {
            // og:image
            $ogImage = $crawler->filter('meta[property="og:image"]');
            if ($ogImage->count() > 0) {
                $images[] = $ogImage->attr('content');
            }

            // Images dalam article/content
            $contentSelectors = ['article', '.post-content', '.article-content', '.content', 'main'];
            foreach ($contentSelectors as $selector) {
                try {
                    $el = $crawler->filter($selector);
                    if ($el->count() > 0) {
                        $el->filter('img')->each(function (Crawler $img) use (&$images, $baseUrl) {
                            $src = $img->attr('src');
                            if (!empty($src)) {
                                // Resolve relative URLs
                                if (strpos($src, 'http') !== 0) {
                                    $src = rtrim($baseUrl, '/') . '/' . ltrim($src, '/');
                                }
                                $images[] = $src;
                            }
                        });
                        if (!empty($images)) break;
                    }
                } catch (\Exception $e) {
                    continue;
                }
            }
        } catch (\Exception $e) {
            // ignore
        }

        return array_unique(array_slice($images, 0, 10));
    }

    /**
     * Error result helper
     */
    protected function errorResult(string $url, string $error): array
    {
        return [
            'url' => $url,
            'title' => '',
            'description' => '',
            'author' => '',
            'publish_date' => '',
            'site_name' => parse_url($url, PHP_URL_HOST) ?? '',
            'content' => '',
            'word_count' => 0,
            'images' => [],
            'success' => false,
            'error' => $error,
        ];
    }
}

File: surfer/services/content_extractor_service.py
import re
import logging
import ipaddress
import socket
from urllib.parse import urlparse
from concurrent.futures import ThreadPoolExecutor, as_completed

logger = logging.getLogger(__name__)

NOISE_SELECTORS = [
    'script', 'style', 'nav', 'footer', 'header',
    '.sidebar', '.advertisement', '.ad', '.ads',
    '.cookie-banner', '.popup', '.modal',
    '.social-share', '.related-posts', '.comments',
    '#sidebar', '#footer', '#header', '#nav',
    '.widget', '.newsletter', '.subscription',
]


class ContentExtractorService:
    @staticmethod
    def _is_safe_url(url):
        """Block internal/private IPs to prevent SSRF."""
        try:
            parsed = urlparse(url)
            if parsed.scheme not in ('http', 'https'):
                return False
            hostname = parsed.hostname
            if not hostname:
                return False
            # Block localhost
            if hostname in ('localhost', '127.0.0.1', '::1', '0.0.0.0'):
                return False
            # Resolve and check IP
            try:
                addr = socket.getaddrinfo(hostname, None, socket.AF_UNSPEC)
                for family, _, _, _, sockaddr in addr:
                    ip = ipaddress.ip_address(sockaddr[0])
                    if ip.is_private or ip.is_loopback or ip.is_link_local or ip.is_reserved or ip.is_multicast:
                        return False
            except (socket.gaierror, ValueError):
                pass
            # Block cloud metadata
            if '169.254.169.254' in hostname or 'metadata.google' in hostname:
                return False
            return True
        except Exception:
            return False

    def extract(self, url):
        if not self._is_safe_url(url):
            return self._error_result(url, 'URL not allowed')
        try:
            from scraper.services.shared_client import fetch
            from scraper.services.rate_limiter import throttle
            throttle(url)
            response = fetch(url)

            if response.status_code != 200:
                return self._error_result(url, f"HTTP {response.status_code}")

            return self._parse_content(response.text, url)
        except Exception as e:
            logger.error(f"Content extraction error for {url}: {e}")
            return self._error_result(url, str(e))

    def extract_multiple(self, urls, max_workers=5):
        results = [None] * len(urls)

        with ThreadPoolExecutor(max_workers=max_workers) as executor:
            future_to_idx = {executor.submit(self.extract, url): i for i, url in enumerate(urls)}
            for future in as_completed(future_to_idx):
                idx = future_to_idx[future]
                try:
                    results[idx] = future.result()
                except Exception as e:
                    results[idx] = self._error_result(urls[idx], str(e))

        return results

    def _parse_content(self, html, url):
        try:
            from lxml import html as lxml_html
            doc = lxml_html.fromstring(html)
        except Exception:
            return self._error_result(url, "Failed to parse HTML")

        # Remove noise elements before extraction
        self._remove_noise(doc)

        title = self._extract_title(doc, html)
        description = self._extract_meta_description(doc, html)
        author = self._extract_author(doc, html)
        publish_date = self._extract_date(doc, html)
        site_name = self._extract_site_name(doc, html, url)
        main_content = self._extract_main_content(doc, html)
        images = self._extract_images(doc, html, url)
        word_count = len(main_content.split()) if main_content else 0

        return {
            'url': url,
            'title': title,
            'description': description,
            'author': author,
            'publish_date': publish_date,
            'site_name': site_name,
            'content': main_content,
            'word_count': word_count,
            'images': images,
            'success': True,
        }

    def _remove_noise(self, doc):
        """Remove noise elements (ads, nav, footer, etc.) from the DOM."""
        for selector in NOISE_SELECTORS:
            try:
                elements = doc.cssselect(selector)
                for el in elements:
                    parent = el.getparent()
                    if parent is not None:
                        parent.remove(el)
            except Exception:
                continue

    def _extract_title(self, doc, html):
        try:
            # og:title
            match = re.search(r'<meta[^>]+property=["\']og:title["\'][^>]+content=["\'](.*?)["\']', html, re.IGNORECASE)
            if match:
                return match.group(1).strip()

            # h1
            try:
                h1 = doc.find('.//h1')
                if h1 is not None and h1.text:
                    return h1.text.strip()
            except Exception:
                pass

            # <title>
            match = re.search(r'<title[^>]*>(.*?)</title>', html, re.DOTALL | re.IGNORECASE)
            if match:
                return re.sub(r'<[^>]*>', '', match.group(1)).strip()
        except Exception:
            pass
        return 'Untitled'

    def _extract_meta_description(self, doc, html):
        try:
            match = re.search(r'<meta[^>]+property=["\']og:description["\'][^>]+content=["\'](.*?)["\']', html, re.IGNORECASE)
            if match:
                return match.group(1).strip()
            match = re.search(r'<meta[^>]+name=["\']description["\'][^>]+content=["\'](.*?)["\']', html, re.IGNORECASE)
            if match:
                return match.group(1).strip()
        except Exception:
            pass
        return ''

    def _extract_author(self, doc, html):
        try:
            for pattern in [
                r'<meta[^>]+name=["\']author["\'][^>]+content=["\'](.*?)["\']',
                r'<meta[^>]+property=["\']article:author["\'][^>]+content=["\'](.*?)["\']',
            ]:
                match = re.search(pattern, html, re.IGNORECASE)
                if match:
                    return match.group(1).strip()

            try:
                el = doc.find('.//*[@itemprop="author"]')
                if el is not None and el.text:
                    return el.text.strip()
            except Exception:
                pass

            for selector in ['.author', '.byline', '.writer']:
                try:
                    els = doc.cssselect(selector)
                    if els and els[0].text:
                        return els[0].text.strip()
                except Exception:
                    continue
        except Exception:
            pass
        return ''

    def _extract_date(self, doc, html):
        try:
            for pattern in [
                r'<meta[^>]+property=["\']article:published_time["\'][^>]+content=["\'](.*?)["\']',
            ]:
                match = re.search(pattern, html, re.IGNORECASE)
                if match:
                    return match.group(1).strip()

            try:
                el = doc.find('.//*[@itemprop="datePublished"]')
                if el is not None:
                    return (el.get('content') or el.text or '').strip()
            except Exception:
                pass

            try:
                time_el = doc.find('.//time')
                if time_el is not None:
                    return (time_el.get('datetime') or time_el.text or '').strip()
            except Exception:
                pass
        except Exception:
            pass
        return ''

    def _extract_site_name(self, doc, html, url):
        try:
            match = re.search(r'<meta[^>]+property=["\']og:site_name["\'][^>]+content=["\'](.*?)["\']', html, re.IGNORECASE)
            if match:
                return match.group(1).strip()
        except Exception:
            pass
        return urlparse(url).hostname or ''

    def _extract_main_content(self, doc, html):
        try:
            # Try <article> first
            try:
                articles = doc.cssselect('article')
                if articles:
                    text = self._get_text(articles[0])
                    if len(text) > 200:
                        return text
            except Exception:
                pass

            # Try content selectors
            for selector in ['.post-content', '.article-content', '.entry-content',
                            '.content', '.main-content', '#content', '#main',
                            '[role="main"]', 'main']:
                try:
                    els = doc.cssselect(selector)
                    if els:
                        text = self._get_text(els[0])
                        if len(text) > 200:
                            return text
                except Exception:
                    continue

            # Fallback: all <p> tags
            try:
                paragraphs = doc.cssselect('p')
                text_parts = []
                for p in paragraphs:
                    p_text = (p.text_content() if hasattr(p, 'text_content') else (p.text or '')).strip()
                    if len(p_text) > 30:
                        text_parts.append(p_text)
                text = '\n\n'.join(text_parts)
                if len(text) > 100:
                    return self._clean_text(text)
            except Exception:
                pass

            # Last fallback: body text
            try:
                body = doc.find('.//body')
                if body is not None:
                    return self._get_text(body)
            except Exception:
                pass
        except Exception as e:
            logger.error(f"Content extraction error: {e}")

        return ''

    def _extract_images(self, doc, html, url):
        images = []
        try:
            match = re.search(r'<meta[^>]+property=["\']og:image["\'][^>]+content=["\'](.*?)["\']', html, re.IGNORECASE)
            if match:
                images.append(match.group(1))

            for selector in ['article', '.post-content', '.article-content', '.content', 'main']:
                try:
                    els = doc.cssselect(selector)
                    if els:
                        for img in els[0].cssselect('img'):
                            src = img.get('src', '')
                            if src:
                                if not src.startswith('http'):
                                    src = url.rstrip('/') + '/' + src.lstrip('/')
                                images.append(src)
                        if images:
                            break
                except Exception:
                    continue
        except Exception:
            pass

        seen = set()
        unique = []
        for img in images:
            if img not in seen:
                seen.add(img)
                unique.append(img)
        return unique[:10]

    def _get_text(self, element):
        try:
            if hasattr(element, 'text_content'):
                text = element.text_content()
            else:
                text = element.text or ''
                for child in element:
                    text += ' ' + self._get_text(child)
            return self._clean_text(text)
        except Exception:
            return ''

    def _clean_text(self, text):
        import html as html_mod
        text = html_mod.unescape(text)
        text = html_mod.unescape(text)
        text = re.sub(r'<[^>]*>', ' ', text)
        text = re.sub(r'\s+', ' ', text)
        text = re.sub(r'\n{3,}', '\n\n', text)
        return text.strip()

    def _error_result(self, url, error):
        return {
            'url': url,
            'title': '',
            'description': '',
            'author': '',
            'publish_date': '',
            'site_name': urlparse(url).hostname or '',
            'content': '',
            'word_count': 0,
            'images': [],
            'success': False,
            'error': error,
        }
